include/wlr/types/wlr_input_method.h

```diff
@@ -0,0 +1,73 @@
+#ifndef WLR_TYPES_WLR_INPUT_METHOD_H
+#define WLR_TYPES_WLR_INPUT_METHOD_H
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <wayland-server.h>
+
+
+struct wlr_input_method_preedit_string {
+	char *text;
+	int32_t cursor_begin;
+	int32_t cursor_end;
+};
+
+struct wlr_input_method_delete_surrounding_text {
+	uint32_t before_length;
+	uint32_t after_length;
+};
+
+struct wlr_input_method_state {
+	struct wlr_input_method_preedit_string preedit;
+	char *commit_text;
+	struct wlr_input_method_delete_surrounding_text delete;
+};
+
+struct wlr_input_method {
+	struct wl_resource *resource;
+
+	struct wlr_input_method_state pending;
+	struct wlr_input_method_state current;
+	bool active; // pending compositor-side state
+	bool client_active; // state known to the client
+	uint32_t current_serial; // received in last commit call
+
+	struct {
+		struct wl_signal commit; // struct wlr_input_method*
+		struct wl_signal destroy; // struct wlr_input_method*
+	} events;
+};
+
+struct wlr_input_method_manager {
+	struct wl_global *input_method_manager;
+	struct wl_resource *resource; // FIXME: turn into a list, each entry per-client
+
+	struct wlr_input_method *input_method; // FIXME: turn into a list, each entry per-seat
```

Comment:
Can you fix this now?

Reply:
The first FIXME pertains to a list of all input methods - since input methods are meant to take over inputs using grabs, does it make sense to allow multiple ones?